test/core/event_engine/posix/posix_event_engine_connect_test.cc

```diff
@@ -0,0 +1,240 @@
+  // be allowed by the kernel before any subsequent connection attempts
+  // become pending indefinitely.
+  while (true) {
+    client_socket = socket(AF_INET6, SOCK_STREAM, 0);
+    setsockopt(client_socket, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
+    // Make fd non-blocking.
+    flags = fcntl(client_socket, F_GETFL, 0);
+    EXPECT_EQ(fcntl(client_socket, F_SETFL, flags | O_NONBLOCK), 0);
+
+    if (connect(client_socket,
+                const_cast<struct sockaddr*>(resolved_addr.address()),
+                resolved_addr.size()) == -1) {
+      if (errno == EINPROGRESS) {
+        struct pollfd pfd;
+        pfd.fd = client_socket;
+        pfd.events = POLLOUT;
+        pfd.revents = 0;
+        int ret = poll(&pfd, 1, 1000);
+        if (ret == -1) {
+          gpr_log(GPR_ERROR, "poll() failed during connect; errno=%d", errno);
+          abort();
+        } else if (ret == 0) {
+          // current connection attempt timed out. It indicates that the
+          // kernel will cause any subsequent connection attempts to
+          // become pending indefinitely.
+          ret_sockets.push_back(client_socket);
+          return ret_sockets;
+        }
+      } else {
+        gpr_log(GPR_ERROR, "Failed to connect to the server (errno=%d)", errno);
+        abort();
+      }
+    }
+    ret_sockets.push_back(client_socket);
+  }
+  return ret_sockets;
+}
+
+}  // namespace
+
+TEST(PosixEventEngineTest, IndefiniteConnectTimeoutOrRstTest) {
+  std::string target_addr = absl::StrCat(
+      "ipv6:[::1]:", std::to_string(grpc_pick_unused_port_or_die()));
+  auto resolved_addr = URIToResolvedAddress(target_addr);
+  std::shared_ptr<EventEngine> posix_ee = std::make_shared<PosixEventEngine>();
+  std::string resolved_addr_str =
+      SockaddrToString(&resolved_addr, true).value();
+  auto sockets = CreateConnectedSockets(resolved_addr);
+  grpc_core::Notification signal;
+  ASSERT_TRUE(sockets.ok());
+  grpc_core::ChannelArgs args;
+  auto quota = grpc_core::ResourceQuota::Default();
+  args = args.Set(GRPC_ARG_RESOURCE_QUOTA, quota);
+  ChannelArgsEndpointConfig config(args);
+  auto memory_quota = absl::make_unique<grpc_core::MemoryQuota>("bar");
+  posix_ee->Connect(
+      [&signal, &resolved_addr_str](
+          absl::StatusOr<std::unique_ptr<EventEngine::Endpoint>> status) {
+        ASSERT_FALSE(status.ok());
+        absl::Status deadline_exceeded_expected_status =
+            absl::CancelledError(absl::StrCat(
+                "Failed to connect to remote host: ", resolved_addr_str,
+                " with error: ",
+                absl::DeadlineExceededError("connect() timed out").ToString()));
+        absl::Status conn_reset_expected_status =
+            absl::CancelledError(absl::StrCat(
+                "Failed to connect to remote host: ", resolved_addr_str,
+                " with error: ",
+                absl::InternalError(absl::StrCat("getsockopt(SO_ERROR): ",
```

Comment:
InternalError usually implies an implementation problem, which ... I don't think this is. https://github.com/abseil/abseil-cpp/blob/85cc4947dbec515868e47a7e4e2b604b21e333cb/absl/status/status.h#L231-L236. That distinction may become important, so we should probably fix the implementation to ensure it returns something more appropriate, maybe FailedPrecondition.

Reply:
Replaced it with FailedPreconditionError